### ISSUE TITLE
Motivation for lmdb-go

### ISSUE THREAD

**Issue by robert-zaremba** [ledgerwatch/lmdb-go#3]
Hi, 
I'm planning to use an LMBD in a go / cosmos project. I know that you did an amazing work with TurboGeth. Could you share please a bit about your motivation for creating this package? Specifically:

Did you consider going [bbolt](https://github.com/etcd-io/bbolt)? It offers a bit more features and data security. BoltDB has a bit worse performance, but not sure if this is a bottleneck. On the other hand it  has more data security features.

**Comment by AlexeyAkhunov** [comment]
Hi Robert,
1. Are you asking why we are using LevelDB (which is what levigo is wrapping) vs LMDB? We need transactionality (ACID or at least ACI) in the database, and LevelDB does not give that. Also, LevelDB is write-optimised, and LMDB (and bolt) are read optimised.
2. Turbo-Geth was using BoltDB for most of its development and we only recently switched to LMDB. Contrary to what it may seem, LMDB actually has more features and more flexible, it just takes time to learn how to use them.

Warning - I did try to simply replace LevelDB with BoltDB in Cosmos SDK, and it got 2x slower. Simply changing the DB won't help much unless you are also prepared to change your data model, which is much harder

**Comment by robert-zaremba** [comment]
@AlexeyAkhunov , thanks for insights!
Cosmos is currently using IAVL+ for the internal storage.
Wow, I wouldn't expect that BoltDB is 2x slower.

What is the performance diff after the switch from BoltDB to LMDB?

**Comment by AlexeyAkhunov** [comment]
>Cosmos is currently using IAVL+ for the internal storage.

Yes, I am aware, I worked with it a tiny bit: https://github.com/AlexeyAkhunov/turbo-geth-archive/tree/turbo-geth-12/avl
:)

Initially when we switched from BoltDB to LMDB, there was no difference. But then we started using features that weren't available in BoltDB, and performance did improve. It is hard to say "how much", because we have variety of workloads in different stages. But in the most challenging stage, I think we reduced the transaction replay time probably by 50-70%.

Again, word of caution - do not expect sudden gains simply from switching the database, you must understand exactly WHY you are switching.
